In [2]:
import sqlite3
import pandas as pd

# Tạo kết nối cơ sở dữ liệu SQLite trong bộ nhớ
conn = sqlite3.connect(':memory:')
cursor = conn.cursor()

# Khởi tạo bảng và chèn dữ liệu
cursor.executescript('''
CREATE TABLE course (
    id INTEGER PRIMARY KEY,
    course_name TEXT
);

CREATE TABLE student (
    student_id INTEGER PRIMARY KEY,
    name TEXT,
    class TEXT,
    course_id INTEGER,
    score REAL
);

INSERT INTO course (id, course_name) VALUES
(12, 'Giai tich'),
(34, 'Thong ke'),
(26, 'Tin hoc');

INSERT INTO student (student_id, name, class, course_id, score) VALUES
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7),
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2),
(3, 'Pham Van Nam', 'Toan Tin', NULL, 7.9),
(4, 'Le Thanh Huyen', 'Toan Tin', 20, 7.2),
(5, 'Vu Quoc Anh', 'May Tinh', 24, 8.0),
(6, 'Dang Thuy Linh', 'May Tinh', 24, 5.5),
(7, 'Bui Tien Dung', 'Kinh Te', 34, 9.2),
(8, 'Ho Ngoc Mai', 'Toan Tin', 20, 8.8),
(9, 'Duong Huu Phuc', 'Kinh Te', NULL, 7.2),
(10, 'Cao Thi Hanh', 'May Tinh', NULL, 7.0);
''')
conn.commit()

# Kiểm tra dữ liệu
display(pd.read_sql_query('SELECT * FROM student;', conn))
display(pd.read_sql_query('SELECT * FROM course;', conn))

,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7
1,2,Tran Thi Lan,Kinh Te,34.0,9.2
2,3,Pham Van Nam,Toan Tin,NaN,7.9
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2
4,5,Vu Quoc Anh,May Tinh,24.0,8.0
5,6,Dang Thuy Linh,May Tinh,24.0,5.5
6,7,Bui Tien Dung,Kinh Te,34.0,9.2
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2
9,10,Cao Thi Hanh,May Tinh,NaN,7.0


,id,course_name
0,12,Giai tich
1,26,Tin hoc
2,34,Thong ke


# Bài 1

In [3]:
# 1. Tích Descartes
df_cartesian = pd.read_sql_query('SELECT * FROM student, course;', conn)
print("1. Tích Descartes:")
display(df_cartesian)

# 2. INNER JOIN
df_inner = pd.read_sql_query('SELECT * FROM student INNER JOIN course ON student.course_id = course.id;', conn)
print("2. INNER JOIN:")
display(df_inner)

# 3. LEFT JOIN
df_left = pd.read_sql_query('SELECT * FROM student LEFT JOIN course ON student.course_id = course.id;', conn)
print("3. LEFT JOIN:")
display(df_left)
# 4 RIGHT
try:
    df_right = pd.read_sql_query('SELECT * FROM student RIGHT JOIN course ON student.course_id = course.id;', conn)
    print("4. RIGHT JOIN:")
    display(df_right)
except Exception as e:
    print("4. RIGHT JOIN bị lỗi do giới hạn của SQLite:", e)
#5 FULL OUTER JOIN
try:
    df_full = pd.read_sql_query('SELECT * FROM student FULL OUTER JOIN course ON student.course_id = course.id;', conn)
    print("5. FULL OUTER JOIN:")
    display(df_full)
except Exception as e:
    print("5. FULL OUTER JOIN bị lỗi do giới hạn của SQLite:", e)

1. Tích Descartes:


,student_id,name,class,course_id,score,id,course_name
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,12,Giai tich
1,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,26,Tin hoc
2,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,34,Thong ke
3,2,Tran Thi Lan,Kinh Te,34.0,9.2,12,Giai tich
4,2,Tran Thi Lan,Kinh Te,34.0,9.2,26,Tin hoc
5,2,Tran Thi Lan,Kinh Te,34.0,9.2,34,Thong ke
6,3,Pham Van Nam,Toan Tin,NaN,7.9,12,Giai tich
7,3,Pham Van Nam,Toan Tin,NaN,7.9,26,Tin hoc
8,3,Pham Van Nam,Toan Tin,NaN,7.9,34,Thong ke
9,4,Le Thanh Huyen,Toan Tin,20.0,7.2,12,Giai tich


2. INNER JOIN:


,student_id,name,class,course_id,score,id,course_name
0,1,Nguyen Minh Hoang,May Tinh,12,6.7,12,Giai tich
1,2,Tran Thi Lan,Kinh Te,34,9.2,34,Thong ke
2,7,Bui Tien Dung,Kinh Te,34,9.2,34,Thong ke


3. LEFT JOIN:


,student_id,name,class,course_id,score,id,course_name
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,12.0,Giai tich
1,2,Tran Thi Lan,Kinh Te,34.0,9.2,34.0,Thong ke
2,3,Pham Van Nam,Toan Tin,NaN,7.9,NaN,None
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2,NaN,None
4,5,Vu Quoc Anh,May Tinh,24.0,8.0,NaN,None
5,6,Dang Thuy Linh,May Tinh,24.0,5.5,NaN,None
6,7,Bui Tien Dung,Kinh Te,34.0,9.2,34.0,Thong ke
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8,NaN,None
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2,NaN,None
9,10,Cao Thi Hanh,May Tinh,NaN,7.0,NaN,None


4. RIGHT JOIN:


,student_id,name,class,course_id,score,id,course_name
0,1.0,Nguyen Minh Hoang,May Tinh,12.0,6.7,12,Giai tich
1,2.0,Tran Thi Lan,Kinh Te,34.0,9.2,34,Thong ke
2,7.0,Bui Tien Dung,Kinh Te,34.0,9.2,34,Thong ke
3,NaN,None,None,NaN,NaN,26,Tin hoc


5. FULL OUTER JOIN:


,student_id,name,class,course_id,score,id,course_name
0,1.0,Nguyen Minh Hoang,May Tinh,12.0,6.7,12.0,Giai tich
1,2.0,Tran Thi Lan,Kinh Te,34.0,9.2,34.0,Thong ke
2,3.0,Pham Van Nam,Toan Tin,NaN,7.9,NaN,None
3,4.0,Le Thanh Huyen,Toan Tin,20.0,7.2,NaN,None
4,5.0,Vu Quoc Anh,May Tinh,24.0,8.0,NaN,None
5,6.0,Dang Thuy Linh,May Tinh,24.0,5.5,NaN,None
6,7.0,Bui Tien Dung,Kinh Te,34.0,9.2,34.0,Thong ke
7,8.0,Ho Ngoc Mai,Toan Tin,20.0,8.8,NaN,None
8,9.0,Duong Huu Phuc,Kinh Te,NaN,7.2,NaN,None
9,10.0,Cao Thi Hanh,May Tinh,NaN,7.0,NaN,None


# KẾT LUẬN CÂU 1:

+ Tích Descartes (CROSS JOIN):

Kết quả: Trả về 30 bản ghi (10 sinh viên x 3 môn học).

Kết luận: Phép toán này ghép mỗi dòng của bảng student với tất cả các dòng của bảng course mà không cần điều kiện. Kết quả này mang tính chất tổ hợp, thường chứa nhiều dữ liệu không có ý nghĩa thực tế (ví dụ: ghép sinh viên đang học Toán với thông tin môn Văn).

+ INNER JOIN:

Kết quả: Chỉ trả về 3 bản ghi (Sinh viên Nguyễn Minh Hoàng khớp mã 12; Trần Thị Lan và Bùi Tiến Dũng khớp mã 34).

Kết luận: INNER JOIN chỉ giữ lại các giao điểm hợp lệ. Những sinh viên chưa đăng ký môn (course_id = NULL) hoặc đăng ký mã môn không tồn tại trong bảng course (mã 20, 24) sẽ bị loại bỏ. Ngược lại, môn "Tin học" (mã 26) chưa có ai đăng ký cũng không được hiển thị.

+ LEFT JOIN:

Kết quả: Trả về toàn bộ 10 bản ghi của bảng student (bảng bên trái).

Kết luận: Phép nối này ưu tiên bảng sinh viên. 3 sinh viên có mã môn hợp lệ sẽ hiển thị kèm tên môn học. 7 sinh viên còn lại (mã môn NULL, 20, 24) vẫn được giữ lại trong danh sách nhưng các cột thông tin về môn học (id, course_name) sẽ bị để trống (NaN/None).

+ RIGHT JOIN:

Kết quả: Trả về thông tin của những sinh viên khớp mã môn, cộng thêm dòng của môn "Tin học" (mã 26).

Kết luận: Ưu tiên bảng course (bảng bên phải). Môn "Tin học" (mã 26) dù chưa có sinh viên nào đăng ký vẫn được giữ lại, các cột thông tin sinh viên tương ứng với môn này sẽ bị bỏ trống (NaN/None). Sinh viên đăng ký mã môn sai hoặc NULL bị loại bỏ.

+ FULL OUTER JOIN:

Kết quả: Là sự kết hợp của LEFT JOIN và RIGHT JOIN.

Kết luận: Giữ lại toàn bộ 10 sinh viên (dù đăng ký sai mã hoặc chưa đăng ký) và giữ lại toàn bộ 3 môn học (kể cả môn Tin học chưa có ai học). Các ô không có dữ liệu tương ứng sẽ được điền khoảng trống (NaN/None).

# CÂU 2

In [10]:
# 2.1 Cập nhật dữ liệu thiếu và loại bỏ dữ liệu sai
cursor.executescript('''
-- Điền mã môn học tồn tại trong bảng course (VD: 26 - Tin học) cho các bạn thiếu
UPDATE student SET course_id = 26 WHERE course_id IS NULL;

-- Xóa các bạn có course_id không nằm trong bảng course
DELETE FROM student WHERE course_id NOT IN (SELECT id FROM course);
''')
conn.commit()

# a. Thống kê theo lớp
df_2a = pd.read_sql_query('''
    SELECT class, COUNT(student_id) AS tong_sv, AVG(score) AS diem_tb
    FROM student GROUP BY class;
''', conn)
print("a. Thống kê theo lớp:")
display(df_2a)

# b. Thống kê theo môn học
df_2b = pd.read_sql_query('''
    SELECT course_id, COUNT(student_id) AS tong_sv, AVG(score) AS diem_tb
    FROM student GROUP BY course_id;
''', conn)
print("b. Thống kê theo môn học:")
display(df_2b)

# c. Phân loại thi đua
df_2c = pd.read_sql_query('''
    SELECT course_id, AVG(score) AS diem_tb,
        CASE 
            WHEN AVG(score) >= 9.0 THEN 'Xuất sắc'
            WHEN AVG(score) >= 6.0 AND AVG(score) <= 8.9 THEN 'Tốt'
            WHEN AVG(score) < 6.0 THEN 'Kém'
        END AS phan_loai
    FROM student GROUP BY course_id;
''', conn)
print("c. Phân loại thi đua theo môn:")
display(df_2c)

a. Thống kê theo lớp:


,class,tong_sv,diem_tb
0,Kinh Te,3,8.533333
1,May Tinh,2,6.850000
2,Toan Tin,1,7.900000


b. Thống kê theo môn học:


,course_id,tong_sv,diem_tb
0,12,1,6.700000
1,26,3,7.366667
2,34,2,9.200000


c. Phân loại thi đua theo môn:


,course_id,diem_tb,phan_loai
0,12,6.700000,Tốt
1,26,7.366667,Tốt
2,34,9.200000,Xuất sắc


# KẾT LUẬN CÂU 2:

 1. Về quá trình chuẩn hóa dữ liệu:

+Đã phát hiện và cập nhật 3 bản ghi bị thiếu mã môn học (course_id là NULL gồm các sinh viên mã 3, 9, 10). Các sinh viên này được gán vào môn "Tin học" (Mã 26).

+Đã phát hiện và loại bỏ 4 bản ghi không hợp lệ do chứa mã môn học không tồn tại trong danh mục course (Mã 20 và 24 gồm các sinh viên mã 4, 5, 6, 8).

Sau khi làm sạch, bảng student còn lại 6 sinh viên hợp lệ để đưa vào thống kê.

2. Về kết quả thống kê và phân loại:

+ a. Thống kê theo lớp học: * Lớp Kinh Te có đông sinh viên hợp lệ nhất (3 bạn) với điểm trung bình cao nhất (khoảng 8.53).

+Lớp May Tinh có 2 sinh viên (điểm TB khoảng 6.85).

+Lớp Toan Tin chỉ còn 1 sinh viên hợp lệ (điểm TB 7.90).

+ b. Thống kê theo môn học:

Môn Tin học (mã 26) thu hút nhiều sinh viên nhất (3 bạn) với điểm trung bình đạt 7.37.

Môn Thống kê (mã 34) có 2 sinh viên nhưng chất lượng rất cao, điểm trung bình đạt 9.20.

Môn Giải tích (mã 12) chỉ có 1 sinh viên theo học (điểm 6.70).

+ c. Phân loại thi đua theo môn học:

Môn Thống kê (mã 34) đạt loại Xuất sắc do điểm trung bình đạt 9.20 (>= 9.0).

Môn Giải tích (mã 12) và môn Tin học (mã 26) cùng đạt loại Tốt do điểm trung bình nằm trong khoảng từ 6.0 đến 8.9. Không có môn nào bị xếp loại Kém.

# BÀI 3

In [11]:
# Xếp hạng sinh viên
query_3 = '''
WITH Ranked AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER(ORDER BY score DESC) as rank_chung,
           RANK() OVER(PARTITION BY class ORDER BY score DESC) as rank_lop,
           RANK() OVER(PARTITION BY course_id ORDER BY score DESC) as rank_mon,
           RANK() OVER(ORDER BY score ASC) as rank_chung_thap
    FROM student
)
SELECT * FROM Ranked;
'''
df_3 = pd.read_sql_query(query_3, conn)
print("Bảng xếp hạng sinh viên:")
display(df_3)

print("Top 3 sinh viên điểm cao nhất (Chung):")
display(df_3[df_3['rank_chung'] <= 3][['student_id', 'name', 'score', 'rank_chung']])

print("Top 3 sinh viên điểm thấp nhất (Chung):")
display(df_3[df_3['rank_chung_thap'] <= 3][['student_id', 'name', 'score', 'rank_chung_thap']])

Bảng xếp hạng sinh viên:


,student_id,name,class,course_id,score,rank_chung,rank_lop,rank_mon,rank_chung_thap
0,2,Tran Thi Lan,Kinh Te,34,9.2,1,1,1,5
1,7,Bui Tien Dung,Kinh Te,34,9.2,1,1,1,5
2,3,Pham Van Nam,Toan Tin,26,7.9,3,1,1,4
3,9,Duong Huu Phuc,Kinh Te,26,7.2,4,3,2,3
4,10,Cao Thi Hanh,May Tinh,26,7.0,5,1,3,2
5,1,Nguyen Minh Hoang,May Tinh,12,6.7,6,2,1,1


Top 3 sinh viên điểm cao nhất (Chung):


,student_id,name,score,rank_chung
0,2,Tran Thi Lan,9.2,1
1,7,Bui Tien Dung,9.2,1
2,3,Pham Van Nam,7.9,3


Top 3 sinh viên điểm thấp nhất (Chung):


,student_id,name,score,rank_chung_thap
3,9,Duong Huu Phuc,7.2,3
4,10,Cao Thi Hanh,7.0,2
5,1,Nguyen Minh Hoang,6.7,1


# KẾT LUẬN CÂU 3: Xếp hạng sinh viên


+ Về phương pháp: Việc sử dụng hàm RANK() thay vì ROW_NUMBER() đảm bảo tính công bằng khi có sinh viên trùng điểm nhau (hiển thị cùng một thứ hạng). Hàm thực hiện đánh giá độc lập mà không làm mất cấu trúc dữ liệu ban đầu.

a. Xếp hạng chung: * Top 3 điểm cao nhất ghi nhận sự đồng hạng 1 của Trần Thị Lan và Bùi Tiến Dũng (cùng 9.2 điểm). Do có hai hạng 1, thứ hạng tiếp theo nhảy sang hạng 3 là Phạm Văn Nam (7.9 điểm).

Top 3 điểm thấp nhất gọi tên Nguyễn Minh Hoàng (hạng 1 từ dưới lên - 6.7 điểm), Cao Thị Hạnh (hạng 2) và Dương Hữu Phúc (hạng 3).

b. Xếp hạng theo Lớp (PARTITION BY class): Bằng cách chia cửa sổ dữ liệu theo lớp, lớp "Kinh Te" thể hiện rõ sự áp đảo với 2 vị trí đồng hạng 1 ở mức điểm 9.2. Trong khi đó, thủ khoa của lớp "May Tinh" chỉ đạt 7.0 (Cao Thị Hạnh) và "Toan Tin" là 7.9 (Phạm Văn Nam).

c. Xếp hạng theo Môn học (PARTITION BY course_id): * Môn Thống kê (mã 34): Lan và Dũng chia nhau vị trí Top 1.

Môn Tin học (mã 26): Nam (hạng 1), Phúc (hạng 2) và Hạnh (hạng 3). Môn học này có sự phân hóa điểm rõ rệt nhất.

# CÂU 4


In [12]:
# Thêm trường graduation_date và tính toán
cursor.executescript('''
ALTER TABLE student ADD COLUMN graduation_date DATETIME;

WITH RankedStudents AS (
    SELECT student_id, RANK() OVER(ORDER BY score DESC) AS rank_val
    FROM student
)
UPDATE student
SET graduation_date = DATE('now', '+' || (
    SELECT rank_val FROM RankedStudents WHERE RankedStudents.student_id = student.student_id
) || ' days');
''')
conn.commit()

# Kiểm tra kết quả
df_4 = pd.read_sql_query('SELECT * FROM student;', conn)
print("Bảng sinh viên sau khi tính thời gian tốt nghiệp:")
display(df_4)

Bảng sinh viên sau khi tính thời gian tốt nghiệp:


,student_id,name,class,course_id,score,graduation_date
0,1,Nguyen Minh Hoang,May Tinh,12,6.7,2026-04-09
1,2,Tran Thi Lan,Kinh Te,34,9.2,2026-04-04
2,3,Pham Van Nam,Toan Tin,26,7.9,2026-04-06
3,7,Bui Tien Dung,Kinh Te,34,9.2,2026-04-04
4,9,Duong Huu Phuc,Kinh Te,26,7.2,2026-04-07
5,10,Cao Thi Hanh,May Tinh,26,7.0,2026-04-08


# KẾT LUẬN CÂU 4: Cập nhật Thời gian Tốt nghiệp


+ Về thao tác: Câu lệnh ALTER TABLE ... ADD COLUMN đã bổ sung thành công trường dữ liệu graduation_date với định dạng thời gian tiêu chuẩn (DATETIME) để lưu trữ chính xác thông tin ngày giờ.


+ Về kết quả tính toán: Hàm thời gian (về mặt logic học thuật là DATE_ADD ) đã tịnh tiến thành công mốc thời gian hiện tại dựa trên số hạng tương ứng của từng sinh viên.

+ Nhận xét dữ liệu: Kết quả cho thấy sự phân bổ ngày tốt nghiệp hoàn toàn khớp với năng lực học tập. Những bạn xuất sắc đạt thứ hạng 1 (Trần Thị Lan, Bùi Tiến Dũng) có thời gian tốt nghiệp sớm nhất (chỉ cộng thêm 1 ngày), trong khi bạn xếp hạng 6 (Nguyễn Minh Hoàng) sẽ có thời gian tốt nghiệp muộn nhất do bị cộng dồn thêm tới 6 ngày.